proto/src/graphstore/v1beta/graphstore.proto

```diff
@@ -0,0 +1,123 @@
+syntax = "proto3";
+
+package graphstore.v1beta;
+
+// for now
+option go_package = "github.com/depscloud/api/v1alpha/store";
+
+import "google/protobuf/any.proto";
+
+enum Encoding {
+    // Data format not specified, defaults to JSON.
+    UNSPECIFIED = 0;
+    // Data is stored as JSON for human readability.
+    JSON = 1;
+    // Data is stored as serialized protocol buffers for efficiency.
+    PROTOCOL_BUFFERS = 2;
+}
+
+// GraphItem represents an element of a graph, serialized to a database row.
+message GraphItem {
+    // The kind represents the type of data stored
+    string kind = 1;
+    // k1 is the first part of the key
+    bytes k1 = 2;
+    // k2 is the second part of the key
+    bytes k2 = 3;
+    // k3 is the third part of the key
+    bytes k3 = 4;
+    // Encoding represents how data is serialized.
+    Encoding encoding = 5;
+    // The encoded data.
+    bytes data = 6;
+}
+
+// Nodes represent entities in a graph.
+message Node {
+    string kind = 1;
+    bytes key = 2;
+    google.protobuf.Any body = 3;
+}
+
+// Edges represent associations between nodes.
+message Edge {
+    string kind = 1;
+    bytes from_key = 2;
+    bytes to_key = 3;
+    bytes key = 4;      // optional
+    google.protobuf.Any body = 5;
+}
+
+// PutRequest encapsulates elements we are inserting into the database.
+message PutRequest {
+    repeated Node nodes = 1;
+    repeated Edge edges = 2;
+}
+
+message PutResponse {}
+
+// DeleteRequest encapsulates elements we are removing from the database.
+message DeleteRequest {
+    repeated Node nodes = 1;
+    repeated Edge edges = 2;
+}
+
+message DeleteResponse {}
+
+// ListRequest encapsulates requirements for listing data from the database.
+message ListRequest {
+    string parent = 1;
+    int32 page_size = 2;
+    string page_token = 3;
+    string kind = 4;
+}
+
+// ListResponse encapsulates listed response data.
+message ListResponse {
+    string next_page_token = 1;
+    repeated Node nodes = 2;
+    repeated Edge edges = 3;
+}
+
+// Neighbor represents a node and all the edges pointing to it.
+message Neighbor {
+    Node node = 1;
+    repeated Edge edges = 2;
+}
+
+// EdgeFilter is used to reduce the considered edges
+message EdgeFilter {
+    string kind = 1;
+    google.protobuf.Any body = 2;
```

Comment:
I'm still working out my feelings about using the `Any` kind. Any does come with an `@type` field which technically does the same thing as `kind`.